pyrmq/consumer.py

```diff
@@ -106,7 +120,67 @@ def __create_connection(self) -> BlockingConnection:
         """
         return BlockingConnection(self.connection_parameters)
 
-    def _consume_message(self, channel, method, properties, data) -> None:
+    def _compute_expiration(self, retry_count: int) -> int:
+        """
+        Computes message expiration time from the retry queue in seconds.
+        """
+        b = self.retry_backoff_base
+        n = self.retry_delay * 1000
+        return b ** (retry_count - 1) * n  # 5, 10, 20, 40, 80
+
+    def _publish_to_retry_queue(
+        self, data: dict, properties, retry_reason: str = None
+    ) -> None:
+        """
+        Publishes message to retry queue with the appropriate metadata in the headers.
+        """
+        attempt = (
+            properties.headers and properties.headers.get("x-attempt", 0) + 1
+        ) or 1
+
+        if attempt > self.max_retries:
+            return
+
+        expiration = self._compute_expiration(attempt)
+        now = datetime.now()
+        next_attempt = now + timedelta(seconds=(expiration // 1000))
+        message_properties = {
+            "expiration": str(expiration),
+            "headers": {
+                **(properties.headers or {}),
+                "x-attempt": attempt,
+                "x-max-attempts": self.max_retries,
+                "x-created-at": properties.headers.get("headers", {}).get(
```

Comment:
properties.headers has headers object inside?

Reply:
Also, properties.headers may not be defined here

Reply:
There's so many existence check. Better if introduce a `headers` variable where we assign `properties.headers or {}`.